### DISCUSSION TITLE
Writing custom platform-specific code in Avalonia

### DISCUSSION THREAD

**Discussion by mail2mhossain** [AvaloniaUI/Avalonia#7448] [Q&A]
Is it possible to write custom platform-specific code in Avalonia?

**Comment by kekekeks** [comment]
Yes, either by `RuntimeInformation.IsOSPlatform` checks of via conditional compilation directives if you use multitargeting.

**Reply by glen-nicol** [reply]
How would one go about using 3rd-party libraries that are platform specific and not .NET?
Maybe more pertinent to Avalonia is how to integrate 3rd party native UI libraries? NativeControlHost?

I realize these questions are probably outside the scope of Avalonia and more of a .NET question on how to interop with Java and swift libraries. Since mobile support is a big selling point of Avalonia 11, it would be nice to have the answer documented here or in the docs even if the answer is "not within scope of this project".

**Reply by maxkatz6** [reply]
> Maybe more pertinent to Avalonia is how to integrate 3rd party native UI libraries? NativeControlHost?

Yes, NativeControlHost.

> I realize these questions are probably outside the scope of Avalonia and more of a .NET question on how to interop with Java and swift libraries.

Yes, pretty much. That's a problem that .NET is solving.
And low level solutions are the same as with MAUI/XF for instance.

Microsoft docs are a bit lacking right now, but at least it's working:
https://learn.microsoft.com/en-us/dotnet/standard/native-interop/
And specifically to Android/iOS
https://github.com/xamarin/xamarin-android (including https://github.com/xamarin/java.interop)
https://github.com/xamarin/xamarin-macios

If control already has a .NET wrappers (mostly the case with build in controls), then it's just a question of writing a simple NativeControlHost. We have ControlCatalog sample in this repo and I am planning to write a separated sample project later that will include mobile.
If control/code doesn't have a .NET wrapper, then it would be more complicated task. And specifically Swift is no go, as there is no interop with .NET as far as I know, only ObjC.

**Reply by alessandrofrancesconi** [reply]
I have the same question as @mail2mhossain , in particular I'm on a cross-platform app for Windows 10 and Android. My Solution is composed by:

- The main shared project (MyApp)
- Desktop project (MyApp.Desktop)
- Android project (MyApp.Android)

Let's say that I want to redirect the user to the Store. There are different ways to do that depending if you are on Windows or Android. Let's take for example:

- Windows-way: `Process.Start("explorer shell:appsfolder\<my_app_id>");` (etc etc....)
- Android-way: `new Intent(Intent.ActionView, Uri.Parse("market://details?id=" + <my_app_id>));` (etc etc....)

(this is a very rough example, but I hope you got the point)

So my goal is to define a `public void OpenStore()` method in both Desktop and Android projects, with each specific implementation. But how to call it from the shared code? When I was on Xamarin, I used `DependencyService`, is there something similar?

**Reply by maxkatz6** [reply]
> When I was on Xamarin, I used DependencyService, is there something similar?

It doesn't matter if Xamarin or not. Just use any preferred IoC container. Now there are a lot of alternatives. Where you basically create an interface something like this:
```
public interface IStoreService
{
    void OpenStoreApp(string appId);
}
```
And register different implementations depending on the platform.
This sample is for WPF, but same IoC library can be used with Avalonia too: https://community.devexpress.com/blogs/wpf/archive/2022/02/07/dependency-injection-in-a-wpf-mvvm-application.aspx

**Reply by alessandrofrancesconi** [reply]
> It doesn't matter if Xamarin or not. Just use any preferred IoC container. Now there are a lot of alternatives. Where you basically create an interface something like this:

Indeed, I've solved using the built-in `ServiceCollection` from Microsoft.Extensions.DependencyInjection, thanks
